Taproot P2TR address

dr-orlovsky opened the pull request:
Part of Taproot project https://github.com/orgs/rust-bitcoin/projects/3 and epic #503. Depends on #617

sgeisler commented:
One thing to keep in mind here: you need the new bech32 version for the `Display` implementation for version>0 segwit addresses. So this is blocked on rust-bitcoin/rust-bech32#48.

dr-orlovsky commented:
@apoelstra what do you  think on refactoring `Address` type given the opportunity by Taproot to do a API breaking change? Why not make it `Copy` and instead of `WitnessProgramm` provide explicit variants for `WPubkeyHash`, `WScriptHash` and `Taproot` with fixed-length arrays? `witness_program()` can be a method returning `&[u8]` then. Each new witness version will probably require address change anyway (like taproot), so we are not creating more problems for the future by doing that.

clarkmoody commented:
I've opened rust-bitcoin/rust-bech32#50 to add Bech32m support to the bech32 library.

sgeisler commented:
> Each new witness version will probably require address change anyway (like taproot)

Ideally that would not be the case. Bech32m was only necessary due to a bug in bech32. BIP173 also states

> Version 0 witness addresses are always 42 or 62 characters, but implementations MUST allow the use of any version. 

Which seems to indicate that we'd still need a "unknown" variant.

clarkmoody commented:
[Here is my take](https://github.com/LNP-BP/rust-bitcoin/compare/taproot/address...clarkmoody:taproot/address) on implementation for Bech32m checksums, based on BIP-0350 and `bech32` 0.8.0.

dr-orlovsky commented:
Thank you, @clarkmoody, your changes are merged

dr-orlovsky commented:
> Which seems to indicate that we'd still need an "unknown" variant.

@sgeisler not necessary. We are talking about addresses, not descriptors; and not each script can be represented by an address (payload). I think it's absolutely fine not to support future unknown address formats and not provide a full witness program for them: ppl needing this will be able to use a special descriptor. In return, we will have a copy'able type like

```rust
#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Debug)]
pub enum AddressPayload {
    PubkeyHash(PubkeyHash),
    ScriptHash(ScriptHash),
    WPubkeyHash(WPubkeyHash),
    WScriptHash(WScriptHash),
    Taproot(bip340::PublicKey),
}
```

and unability to parse the address will be denoted by `Option::None`, as it is happens today for custom script pubkeys.

sgeisler commented:
Imo "*[…] but implementations MUST allow the use of any version.*" is intended to provide forwards compatibility. We don't need to understand a witness script to send money to it and future users shouldn't have to upgrade to send money to version >=2 witness addresses. That upgrading is necessary for sending funds to Taproot addresses is a very unfortunate mishap and not the norm.

dr-orlovsky commented:
However you are sending funds to a scriptPubkey, not the address. So Bech32m have not broken the funds which were stent to witness v1 outputs, and not supporting _parsing_ of an address does not implies preventing users from sending funds wherever they want to. Anyway, it's not that important so I'm fine if it is not needed in the library.

apoelstra commented:
We absolutely should support versions >1. The whole point of bech32(m) is that you can support future versions.

Users send money to addresses, not scriptPubKeys.

clarkmoody commented:
The main bitcoin repo has an [enum variant](https://github.com/bitcoin/bitcoin/blob/bd6af53e1f8ec9d25cedf0bf36c98b99a8d88774/src/script/standard.h#L120-L132) for unknown witness version that it uses when classifying outputs. 

```c++
enum class TxoutType {
    NONSTANDARD,
    // 'standard' transaction types:
    PUBKEY,
    PUBKEYHASH,
    SCRIPTHASH,
    MULTISIG,
    NULL_DATA, //!< unspendable OP_RETURN script that carries data
    WITNESS_V0_SCRIPTHASH,
    WITNESS_V0_KEYHASH,
    WITNESS_V1_TAPROOT,
    WITNESS_UNKNOWN, //!< Only for Witness versions not already defined above
};
```

stevenroose reviewed (dismissed):
LGTM it seems that this version does support segwit versions > 1, right?

I would suggest a follow-up PR to add an extra `AddressType::UnknownSegwitVersion(SegwitVersion)` or something.

dr-orlovsky commented:
> it seems that this version does support segwit versions > 1, right?

in which sense? We have enum values for all allowed versions >1

sgeisler reviewed (dismissed):
utACK, looks good from a high level, would be nice if someone (possibly myself, but later) could take a more detailed look.

sgeisler commented:
> in which sense?

That we can a) parse and format and b) send transactions to addresses of unknown witness version.

> We have enum values for all allowed versions >1

Yes, this makes it possible.

sanket1729 reviewed (changes requested):
Some nits and a bug where we don't check program length for taproot witness programs. 

Also added another point for discussion about having a different type for tweaked keys. Curious to see feedback from other reviewers on it.

stevenroose commented on src/util/address.rs:
```diff
@@ -148,6 +152,171 @@ impl FromStr for AddressType {
+    /// NB: this is not the same as an integer representation of the opcode signifying witness
+    /// version in bitcoin script. Thus, there is no function to directly convert witness version
+    /// into a byte since the conversion requires context (bitcoin script or just a version number)
+    pub fn into_version_no(self) -> u8 {
```
Yeah +1 on the naming. `no` for number if definitely opinionated and I think in the codebase we use `nb` or `num` but never `no`. But something shorter like `into_ver` seems good enough, indeed.

stevenroose commented on src/util/address.rs:
```diff
@@ -474,6 +479,17 @@ impl Address {
         }
     }
 
+    /// Create a pay to taproot address
+    pub fn p2tr(taptweaked_key: schnorrsig::PublicKey, network: Network) -> Address {
+        Address {
+            network: network,
+            payload: Payload::WitnessProgram {
+                version: WitnessVersion::V1,
+                program: taptweaked_key.serialize().to_vec()
+            }
+        }
+    }
```
Hmm, I'm not sure from a high-level interface level it makes sense to distinguish tweaked keys from regular ones. In principle they are the same, right? It's just functionally that they are different. So from an outside observer like here, we can never know if a key is tweaked or not. A single-user wallet f.e. has no downside of just using single keys for taproot outputs, AFAIU.

sanket1729 replied:
The current discussion on descriptor `tr(K)` https://github.com/bitcoin/bitcoin/pull/21365 uses K as internal key. `tr(P)` translates to output key P+H_TapTweak(P)*G.

sgeisler replied:
A new type could be made quite ergonomic by implementing a `tweak(&self) -> TweakedPubKey` function for `schnorrsig::PublicKey` (possibly via a `Tweakable` trait in `bitcoin` if `secp256k1` should stay slim). This function would output the "standard" p2pk tweak. Maybe later there would be a second function to tweak based on a script/taproot node.

dr-orlovsky replied:
I think tweaks are important! But cant fit this PR, so separate PR is needed (also addressing non-taproot tweaks)

sanket1729 replied:
Considering that descriptors use the internal key in them, I can forsee a confusion if we offer this API. Should the user give an internal key or the tweaked key?

We can have this discussion later, I will ACK this PR, but should definitely address this before release.

dr-orlovsky commented on src/util/address.rs:
```diff
@@ -148,6 +152,171 @@ impl FromStr for AddressType {
+    /// Takes bitcoin Script opcode and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_opcode(opcode: opcodes::All) -> Result<Self, Error> {
+        match opcode.into_u8() {
+            0 => Ok(WitnessVersion::V0),
+            ver if ver >= opcodes::all::OP_PUSHNUM_1.into_u8() && ver <= opcodes::all::OP_PUSHNUM_16.into_u8() =>
+                WitnessVersion::from_version_no(ver - opcodes::all::OP_PUSHNUM_1.into_u8() + 1),
+            _ => Err(Error::UnparsableWitnessVersion)
+        }
```
Not sure I understand what you mean. Do you also propose to change opcodes api in this PR? Afraid it will create more discussions and will delay merging of this one

sanket1729 replied:
I mean. This can be 
```
match opcode.classify() {
    Class::PushNum(ver) if ver > 0 => ,
    Class::PushBytes(0) =>
    _ => Err
```

dr-orlovsky replied:
After careful consideration I think that using byte value is more robust, since it does not rely on the correctness of the classification, which is not part of the consensus (and in BIP witness is not defined via classification of opcodes)

dr-orlovsky commented on src/util/address.rs:
```diff
@@ -491,6 +507,7 @@ impl Address {
                         32 => Some(AddressType::P2wsh),
                         _ => None,
                     },
+                    WitnessVersion::V1 => Some(AddressType::P2tr),
```
Yes, but the length is checked during the struct init. We do length check above only because we need to distinguish two addresses

sanket1729 replied:
No, we do the check above because Segwit requires program len to be of size 32 or 20, otherwise we must error. BIP341 specifically mentions 
```
 The following rules only apply when such an output is being spent. Any other outputs, including version 1 outputs with lengths other than 32 bytes, or P2SH-wrapped version 1 outputs[3], remain unencumbered.
```
So, presently this is a bug as this will mark other programs as taproot when they are not.

sanket1729 commented:
@dr-orlovsky if your concern is waiting for 2 reviewer ACKs again, we can also create a follow-up PR with the changes.

dr-orlovsky commented:
Thanks @sanket1729, I will add commits to this PR. Just need more time to finish other stuff before switching to this one

dr-orlovsky commented:
@sanket1729 addressed all of your suggestions in the new commit, aside from the first one (for which I gave rationale why I think the current code is better).

@stevenroose, @sgeisler now this will be pending your re-ACK

sgeisler commented on src/util/address.rs:
```diff
@@ -168,23 +335,17 @@ impl Payload {
-            Payload::PubkeyHash(PubkeyHash::from_slice(&script.as_bytes()[3..23]).unwrap())
+            let mut hash_inner = [0u8; 20];
+            hash_inner.copy_from_slice(&script.as_bytes()[3..23]);
+            Payload::PubkeyHash(PubkeyHash::from_inner(hash_inner))
```
The new and old code seem equivalent, what's the reason for the change?

sanket1729 replied:
Avoids the `unwrap()` I guess. I am indifferent, best left to the code style of the author

sgeisler reviewed (approved):
tACK 7c32c7fc5c9e013cf893fdbbe43665e76d2b240c

<details>
  <summary>Test log:</summary>

```
Apr 06 17:26:33.766  INFO testinator: Installing rust toolchain 'nightly'
Apr 06 17:26:36.389  INFO testinator: Installing rust toolchain 'stable'
Apr 06 17:26:36.643  INFO testinator: Installing rust toolchain '1.29.0'
Apr 06 17:26:36.873  INFO testinator: Preparing environment for rust nightly tests (9 configurations)
Apr 06 17:26:36.874  INFO testinator: Preparing environment for rust stable tests (8 configurations)
Apr 06 17:26:36.874  INFO testinator: Preparing environment for rust 1.29.0 tests (8 configurations)
Apr 06 17:27:44.850  INFO testinator: Running rust nightly tests in /tmp/rust-bitcoin-nightly.q0GCXPPVY41z/rust-bitcoin
Apr 06 17:27:44.854  INFO testinator: Running rust 1.29.0 tests in /tmp/rust-bitcoin-1.29.0.fJk1f8F3peCL/rust-bitcoin
Apr 06 17:27:44.854 DEBUG testinator: Generating lock file with rust=1.29.0
Apr 06 17:27:44.858  INFO testinator: Running rust stable tests in /tmp/rust-bitcoin-stable.ChE2yLlxWOgK/rust-bitcoin
Apr 06 17:27:46.656 DEBUG testinator: Pinning cc to 1.0.41
Apr 06 17:27:46.839 DEBUG testinator: Pinning serde to 1.0.98
Apr 06 17:27:47.030 DEBUG testinator: Pinning serde_derive to 1.0.98
Apr 06 17:27:47.305 DEBUG testinator: Pinning byteorder to 1.3.4
Apr 06 17:28:28.440  INFO testinator: Test rust=nightly, features=[secp-recovery] succeeded!
Apr 06 17:28:37.784  INFO testinator: Test rust=stable, features=[secp-recovery] succeeded!
Apr 06 17:29:03.816  INFO testinator: Test rust=1.29.0, features=[secp-recovery] succeeded!
Apr 06 17:29:13.877  INFO testinator: Test rust=nightly, features=[use-serde] succeeded!
Apr 06 17:29:33.909  INFO testinator: Test rust=stable, features=[use-serde] succeeded!
Apr 06 17:29:35.676  INFO testinator: Test rust=nightly, features=[base64] succeeded!
Apr 06 17:30:05.646  INFO testinator: Test rust=stable, features=[base64] succeeded!
Apr 06 17:30:08.549  INFO testinator: Test rust=nightly, features=[secp-lowmemory] succeeded!
Apr 06 17:30:29.381  INFO testinator: Test rust=nightly, features=[rand] succeeded!
Apr 06 17:30:36.467  INFO testinator: Test rust=stable, features=[secp-lowmemory] succeeded!
Apr 06 17:30:51.175  INFO testinator: Test rust=1.29.0, features=[use-serde] succeeded!
Apr 06 17:30:52.423  INFO testinator: Test rust=nightly, features=[unstable] succeeded!
Apr 06 17:31:09.924  INFO testinator: Test rust=stable, features=[rand] succeeded!
Apr 06 17:31:42.735  INFO testinator: Test rust=nightly, features=[bitcoinconsensus] succeeded!
Apr 06 17:32:01.278  INFO testinator: Test rust=1.29.0, features=[base64] succeeded!
Apr 06 17:32:07.630  INFO testinator: Test rust=nightly, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,unstable,bitcoinconsensus] succeeded!                                                                                                  
Apr 06 17:32:12.992  INFO testinator: Test rust=stable, features=[bitcoinconsensus] succeeded!
Apr 06 17:32:48.787  INFO testinator: Test rust=nightly, features=[] succeeded!
Apr 06 17:32:57.434  INFO testinator: Test rust=1.29.0, features=[secp-lowmemory] succeeded!
Apr 06 17:33:10.115  INFO testinator: Test rust=stable, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,bitcoinconsensus] succeeded!                                                                                                            
Apr 06 17:33:51.636  INFO testinator: Test rust=1.29.0, features=[rand] succeeded!
Apr 06 17:33:51.638  INFO testinator: Test rust=stable, features=[] succeeded!
Apr 06 17:34:40.626  INFO testinator: Test rust=1.29.0, features=[bitcoinconsensus] succeeded!
Apr 06 17:35:28.786  INFO testinator: Test rust=1.29.0, features=[secp-recovery,use-serde,base64,secp-lowmemory,rand,bitcoinconsensus] succeeded!                                                                                                            
Apr 06 17:36:10.203  INFO testinator: Test rust=1.29.0, features=[] succeeded!
Apr 06 17:36:14.164  INFO testinator: Fuzzing deserialize_script
Apr 06 17:37:25.152  INFO testinator: Successfully fuzzed deserialize_script
Apr 06 17:37:25.152  INFO testinator: Fuzzing uint128_fuzz
Apr 06 17:38:27.100  INFO testinator: Successfully fuzzed uint128_fuzz
Apr 06 17:38:27.100  INFO testinator: Fuzzing deserialize_amount
Apr 06 17:39:28.101  INFO testinator: Successfully fuzzed deserialize_amount
Apr 06 17:39:28.101  INFO testinator: Fuzzing deserialize_transaction
Apr 06 17:40:30.258  INFO testinator: Successfully fuzzed deserialize_transaction
Apr 06 17:40:30.258  INFO testinator: Fuzzing deser_net_msg
Apr 06 17:41:33.076  INFO testinator: Successfully fuzzed deser_net_msg
Apr 06 17:41:33.076  INFO testinator: Fuzzing deserialize_address
Apr 06 17:42:34.071  INFO testinator: Successfully fuzzed deserialize_address
Apr 06 17:42:34.071  INFO testinator: Fuzzing deserialize_block
Apr 06 17:43:36.235  INFO testinator: Successfully fuzzed deserialize_block
Apr 06 17:43:36.235  INFO testinator: Fuzzing outpoint_string
Apr 06 17:44:38.151  INFO testinator: Successfully fuzzed outpoint_string
Apr 06 17:44:38.151  INFO testinator: Fuzzing deserialize_psbt
Apr 06 17:45:41.146  INFO testinator: Successfully fuzzed deserialize_psbt
```
</details>

sanket1729 commented on src/util/address.rs:
```diff
@@ -502,7 +477,7 @@ impl Address {
 
     /// Check whether or not the address is following Bitcoin
-    /// standardness rules.
+    /// standard rules.
```
nit (non-merge blocking, not necessary to fix) :

I think standardness is the more used term.

stevenroose replied:
ack

sanket1729 reviewed (dismissed):
tACK 7c32c7fc5c9e013cf893fdbbe43665e76d2b240c.

RCasatta commented on src/util/address.rs:
```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+    }
+
+    /// Determine the checksum variant. See BIP-0350 for specification.
+    pub fn bech32_variant(&self) -> bech32::Variant {
```
feel free to ignore since it's a late comment but I was wondering if it has been considered as `impl Into<bech32::Variant> for WitnessVersion` ?

Kixunil replied:
I guess `From` rather than `Into` but it seems to make sense.

RCasatta replied:
yeah, sorry I meant `impl From<WitnessVersion> for bech32::Variant`

dr-orlovsky replied:
Added in 747d97cdd6e2d66fa10a8c0cf6bac2e1d06dadde

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+    pub fn bech32_variant(&self) -> bech32::Variant {
+        match self {
+            WitnessVersion::V0 => bech32::Variant::Bech32,
+            _ => bech32::Variant::Bech32m,
```
Exhaustive matching is generally better but I guess it's fine here because new fields will never be added? (would break consensus otherwise)

dr-orlovsky replied:
Yes, you can't add variants to `WitnessVersion`

Kixunil commented on src/blockdata/script.rs:
```diff
@@ -257,23 +258,18 @@ impl Script {
+    pub fn new_witness_program(ver: WitnessVersion, program: &[u8]) -> Script {
         Builder::new()
-            .push_opcode(verop.into())
+            .push_opcode(ver.into())
```
Why not `version`? I believe it'd be more readable and doesn't remind people of one misguided/malicious individual...

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let ver = u8::from_str(s).map_err(|_| Error::UnparsableWitnessVersion)?;
```
This loses information from the inner error. :(

I strongly believe errors should not discard valuable information like this. The variant should contain the inner error.

sgeisler replied:
This would require another error variant, but it might be the right thing to do depending on the use case (I'm not sure why `FromStr` and `Display` even exist for a witness version, it's something purely internal imo, but it didn't seem to hurt)

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+            0 => Ok(WitnessVersion::V0),
+            ver if ver >= opcodes::all::OP_PUSHNUM_1.into_u8() && ver <= opcodes::all::OP_PUSHNUM_16.into_u8() =>
+                WitnessVersion::from_ver_u8(ver - opcodes::all::OP_PUSHNUM_1.into_u8() + 1),
+            _ => Err(Error::UnparsableWitnessVersion)
```
Why not `InvalidWitnessVersion`?

sgeisler replied:
To my understanding `InvalidWitnessVersion` s for out of bounds numerical values and `UnparsableWitnessVersion` for things that aren't even valid numerical representations in this context (like all other op codes).

Kixunil replied:
Right, that makes sense.

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+        ::bech32::u5::try_from_u8(ver.into_ver_u8()).expect("WitnessVersion must be 0..=16")
+    }
+}
+
```
Maybe conersion to `u8` would be useful as well?

dr-orlovsky replied:
We can't have it since it should be a failing conversion and with rustc 0.29 we do not have `TryFrom`. To convert `WitnessVersion` to `u8` we already have a dedicated function

stevenroose commented on src/util/address.rs:
```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+    pub fn from_instruction(instruction: Instruction) -> Result<Self, Error> {
+        match instruction {
+            Instruction::Op(op) => WitnessVersion::from_opcode(op),
+            Instruction::PushBytes(bytes) if bytes.len() == 0 => Ok(WitnessVersion::V0),
```
Heh? I don't know this special rule, could you perhaps elaborate?

dr-orlovsky replied:
This actually corresponds to `0` as the first script byte – that is how it is presented at instruction level

stevenroose commented:
utACK 7c32c7f - is there a reason we're holding off of merging?

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+    /// outputs
+    V0,
+    /// Forthcoming second version of Witness Program, which (most probably)
+    /// will be used for Taproot
```
I think waiting for activation before it's presented as a fact is a good thing to do.

dr-orlovsky commented:
@stevenroose we are mainly constrained by 0.26.1 release since this PR is a breaking one and must be merged after it.

Otherwise, I will be happy to address all nits, aside from the fact that this will cancel three existing ACKs :( Not sure what would be the best thing to do

dr-orlovsky commented:
Addressed all nits with a separate commit 6155699b82101ef50be29d5ef65641166852663c, will ask everybody to re-ACK

Kixunil commented on src/util/address.rs:
```diff
@@ -64,7 +65,9 @@ pub enum Error {
-    UnparsableWitnessVersion,
+    UnparsableWitnessVersion(ParseIntError),
+    /// Bitcoin script opcode does not match any known witness version, the script if malformed
+    MalformedWitnessVersion,
```
Would it make sense to report the opcode or (probably more importantly) position within script?

dr-orlovsky replied:
Not trivial, since this error is also produced by instructions, not only opcode. And the position is always first byte, according to SegWit

Kixunil commented on src/util/address.rs:
```diff
@@ -81,7 +84,8 @@ impl fmt::Display for Error {
             Error::InvalidBech32Variant => write!(f, "invalid bech32 checksum variant"),
             Error::InvalidWitnessVersion(v) => write!(f, "invalid witness script version: {}", v),
-            Error::UnparsableWitnessVersion => f.write_str("Incorrect format of a witness version byte"),
+            Error::UnparsableWitnessVersion(ref e) => write!(f, "Incorrect format of a witness version byte: {}", e),
```
The error types that follow `std::error::Error` should not display their sources and leave it to the callers. That way error sources can be formatted differently. E.g. `cargo` separates them by newlines as opposed to colons.

Note that this only applies to sources that themselves implement errors and return `Some(inner)`. If error carries some additional context (paths, strings...) then this context should be displayed.

dr-orlovsky replied:
Well, everywhere throughout the library it is done this way (you can also check the lines above), so changing this method should be a topic for the separate PR

dr-orlovsky replied:
Fixed in 6848ea43aa53e903323408bfe78bec342dc9efa4

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +159,167 @@ impl FromStr for AddressType {
+    /// NB: this is not the same as an integer representation of the opcode signifying witness
+    /// version in bitcoin script. Thus, there is no function to directly convert witness version
+    /// into a byte since the conversion requires context (bitcoin script or just a version number)
+    pub fn into_ver_u8(self) -> u8 {
```
s/ver/version/?

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +159,167 @@ impl FromStr for AddressType {
+    /// Takes byte value and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_ver_u8(value: u8) -> Result<Self, Error> {
```
s/ver/version/?

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +159,167 @@ impl FromStr for AddressType {
+    pub fn from_opcode(opcode: opcodes::All) -> Result<Self, Error> {
+        match opcode.into_u8() {
+            0 => Ok(WitnessVersion::V0),
+            ver if ver >= opcodes::all::OP_PUSHNUM_1.into_u8() && ver <= opcodes::all::OP_PUSHNUM_16.into_u8() =>
```
s/ver/version/?

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +159,167 @@ impl FromStr for AddressType {
+impl From<WitnessVersion> for opcodes::All {
+    /// Converts `WitnessVersion` instance into corresponding Bitcoin script
+    /// opcode (`OP_0`..`OP_16`)
+    fn from(ver: WitnessVersion) -> opcodes::All {
```
s/ver/version/?

Kixunil commented on src/util/address.rs:
```diff
@@ -143,11 +159,167 @@ impl FromStr for AddressType {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let ver = u8::from_str(s).map_err(|err| Error::UnparsableWitnessVersion(err))?;
```
s/ver/version/?

Kixunil commented on src/util/address.rs:
```diff
@@ -319,20 +496,21 @@ impl Address {
                 // BIP-141 p2wpkh or p2wsh addresses.
-                match ver.to_u8() {
-                    0 => match prog.len() {
+                match ver {
```
s/ver/version/?

Kixunil reviewed (commented):
Found a few more instances of `ver` instead of `version`. Other than that it looks good.

sgeisler commented:
Since you changed the other occurrences of `ver` I assume you want to fix these too @dr-orlovsky, so I'll wait with another review. Hopefully it will be the final one then :laughing:

dr-orlovsky commented:
I would not say that I see any good rationale  behind this bikeshedding, but I will change the names, ok

sgeisler commented:
> I would not say that I see any good rationale behind this bikeshedding, but I will change the names, ok

Me neither, I just wanted to avoid getting my review invalidated again xD

sgeisler commented on src/util/address.rs:
```diff
@@ -81,7 +84,8 @@ impl fmt::Display for Error {
             Error::InvalidBech32Variant => write!(f, "invalid bech32 checksum variant"),
             Error::InvalidWitnessVersion(v) => write!(f, "invalid witness script version: {}", v),
-            Error::UnparsableWitnessVersion => f.write_str("Incorrect format of a witness version byte"),
+            Error::UnparsableWitnessVersion(ref e) => write!(f, "Incorrect format of a witness version byte: {}", e),
```
This error occurs when parsing a string, idk if referring to a byte is really the right thing to do. Super minor nit though.

dr-orlovsky replied:
Fixed in 6848ea43aa53e903323408bfe78bec342dc9efa4

sgeisler reviewed (approved):
utACK 477983372e09bdc91862a7788d71f96c53e521f7

dr-orlovsky commented:
Converting to draft since pending merge of #617 (basing on it and requiring rebase once it will got merged)

sanket1729 commented on src/util/address.rs:
```diff
@@ -655,6 +666,11 @@ impl FromStr for Address {
 
+            // Encoding check
+            if version.bech32_variant() != variant {
+                return Err(Error::InvalidBech32Variant);
```
nit: This could provide more context about the error by also providing version and variant.

dr-orlovsky replied:
Fixed in 6848ea43aa53e903323408bfe78bec342dc9efa4

sanket1729 reviewed (dismissed):
ACK 800261d78b0fa81f6e6619ee35561769c3ac1da2 . The module level documentation can also be amended with one-two line description bech32/bech32m.

apoelstra commented:
Ok, merged #617, so this can be un-drafted.

dr-orlovsky commented:
@apoelstra rebased, updated, addressed comments & ready for review

Kixunil commented on src/util/address.rs:
```diff
@@ -59,6 +60,13 @@ pub enum Error {
+        expected: bech32::Variant,
+        /// The actual Bech32 variant encoded in the address representation
+        found: bech32::Variant
+    },
```
Awesome error reporting here!

Kixunil commented on src/util/address.rs:
```diff
@@ -76,11 +84,12 @@ pub enum Error {
-            Error::Base58(ref e) => write!(f, "base58: {}", e),
-            Error::Bech32(ref e) => write!(f, "bech32: {}", e),
+            Error::Base58(_) => write!(f, "base58 address encoding error"),
+            Error::Bech32(_) => write!(f, "bech32 address encoding error"),
```
`f.push_str("...")` would be a bit more performant but almost certainly doesn't matter. Feel free to ignore.

dr-orlovsky replied:
Corrected

Kixunil commented on src/util/address.rs:
```diff
@@ -499,6 +538,7 @@ impl Address {
                         32 => Some(AddressType::P2wsh),
                         _ => None,
                     },
+                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
```
I guess better for a different PR but I'd like to see this `fn` return `Result` with more information in `Err` - in this case invalid length.

dr-orlovsky replied:
You are right, and actually fixing that led me to discover another hidden error: some BIP350 test vecs are correct as Bech32m, but are incorrect as address representation. So I have added new error types, required checks etc in the last separate commit to this PR

Kixunil reviewed (approved):
LGTM

Kixunil commented on src/util/address.rs:
```diff
@@ -523,23 +531,23 @@ impl Address {
-                    _ => None,
+                    WitnessVersion::V1 if prog.len() == 32 => Ok(AddressType::P2tr),
+                    version => Err(Error::UnsupportedWitnessVersion(version)),
                 }
```
`match (version, prog.len())` would look a bit nicer but not a big deal.

But thinking about it isn't `Address` supposed to guarantee that lengths are always valid? This is checked at parsing and I don't seen an alternative unchecked constructor. If I'm right then these checks should be unnecessary. If not, then maybe it'd be better to have (private) `fn check_program_len(self, len: usize) -> Result<(), Error>` method on `WitnessVersion` and use it in both places to DRY.

dr-orlovsky replied:
I was thinking about this myself. However, one day somebody may introduce new methods to create an address without checks and without this error checking we will get problems. So I think it is better to have an unreachable error checking code rather than assume everything should be fine due to the code in other external methods and unwrap/expect.

Kixunil replied:
Right, maybe a more relevant question is if we should commit into address having this invariant.

dr-orlovsky replied:
I think this is outside of the scope of this PR and may be controversial on how to achieve this

Kixunil commented on src/util/address.rs:
```diff
@@ -77,29 +77,37 @@ pub enum Error {
             ),
-            Error::UncompressedPubkey => write!(f,
+            Error::InvalidTaprootProgramLength(l) => write!(f,
+                "a v1 witness program (taproot) must be either of length 20 or 32 bytes: length={}", l,
```
This `l` looks a bit like `1` so would be nice to use `len` instead.

sanket1729 replied:
Same length error here. v1 witness programs are always 32 bytes

Kixunil replied:
The len is what was given, not what is expected.

Kixunil commented on src/util/address.rs:
```diff
@@ -604,7 +612,7 @@ impl Address {
-            Some(AddressType::P2pkh) | Some(AddressType::P2sh) => true,
+            Ok(AddressType::P2pkh) | Ok(AddressType::P2sh) => true,
             _ => false
         };
```
Would it be useful to have (public) `is_legacy(&self)` method?

dr-orlovsky replied:
How we will define a "legacy"? Just P2PK(H)/P2SH or SegWit v0 as well?

Kixunil replied:
Same as this code: not segwit

Although now that you bring it up, perhaps `is_segwit()` would be a more readable name. But more importantly, I'm not sure if this is correct if there's an error: https://github.com/rust-bitcoin/rust-bitcoin/pull/563/files#diff-c55d0186f3d5990b5e223c1a8aa1bbe60988201c452dd13b3324e892bc33119eR624 If it's not then `is_segwit()` instead of `!is_legacy` would seem correct.

dr-orlovsky replied:
This code was not introduced by this PR and not related to it, so probably we need to file an issue and do a proper investigation

Kixunil reviewed (approved):
I had a few more stylistic ideas but the code looks correct.

sanket1729 commented on src/util/address.rs:
```diff
@@ -714,9 +722,15 @@ impl FromStr for Address {
+                (WitnessVersion::V0, len) if len != 20 && len != 32 => {
+                    return Err(Error::InvalidSegwitV0ProgramLength(len))
+                },
+                (WitnessVersion::V1, len) if len != 32 => {
```
> Any other outputs, including version 1 outputs with lengths other than 32 bytes, or P2SH-wrapped version 1 outputs[3], remain unencumbered.

As far as I can recollect, this is not correct. We can have (I hope we never do) a new `(WitnessVersion::V1, 33)` in the future.

sanket1729 replied:
I really want to return this error here, but I think that would be wrong.

dr-orlovsky replied:
No, since non-32 byte V1 will be a hard fork

sanket1729 replied:
> Any other outputs, including version 1 outputs with lengths other than 32 bytes, or P2SH-wrapped version 1 outputs[3], remain unencumbered.

Sorry, I should have highlighted that this statement is from BIP341. And I have verified with the implementation, that non-32 byte V1 won't be a hard fork. I hope never to see that soft fork, but it is what it is.

dr-orlovsky replied:
Wow, thank you for spotting that! Are you aware of the rationale for a such weird thing?

Kixunil replied:
I guess it's left for possible future upgrades in case we run out of version numbers or something.

sanket1729 commented on src/util/address.rs:
```diff
@@ -523,23 +531,23 @@ impl Address {
-                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
-                    _ => None,
+                    WitnessVersion::V1 if prog.len() == 32 => Ok(AddressType::P2tr),
+                    version => Err(Error::UnsupportedWitnessVersion(version)),
```
nit: Can we also add len field over here? Because this will show an error as `UnsupportedWitnessVersion(1)` or `0` which might be confusing

sanket1729 reviewed (commented):
Looks great overall. Left a few comments

sanket1729 commented on src/util/address.rs:
```diff
@@ -523,23 +533,23 @@ impl Address {
                     },
-                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
-                    _ => None,
+                    WitnessVersion::V1 => Ok(AddressType::P2tr),
```
we still need the `if prog.len() == 32` check

dr-orlovsky replied:
Sorry, I am not getting why if you just pointed out that this is not a part of the spec

dr-orlovsky replied:
Ok, got it. If we will have non-32-byte V1 it will be _non-taproot_ V1

sanket1729 reviewed (dismissed):
ACK  3d23ac71866581af4b6558780dbab684baacb52d

sgeisler commented on src/util/address.rs:
```diff
@@ -907,8 +954,8 @@ mod tests {
         for vector in &valid_vectors {
-            let addr: Address = vector.0.parse().unwrap();
-            assert_eq!(&addr.script_pubkey().as_bytes().to_hex(), vector.1);
+            let addr: Address = vector.0.parse().expect(&format!("error in test vector {}", vector.0));
```
TIL: non-static strings are allowed in `expect`. Although I'd avoid this in production code as it always has to do the string interpolation, not only in the error case. Ok for tests though imo.

sgeisler reviewed (approved):
utACK 3d23ac71866581af4b6558780dbab684baacb52d

I left a few nits, but none were blocking in the end imo. I could understand if you don't want to invalidate reviews. Haven't had much time for it recently :( hope it gets better

dr-orlovsky commented on src/util/address.rs:
```diff
@@ -71,26 +79,37 @@ pub enum Error {
-            Error::Base58(ref e) => write!(f, "base58: {}", e),
-            Error::Bech32(ref e) => write!(f, "bech32: {}", e),
-            Error::EmptyBech32Payload => write!(f, "the bech32 payload was empty"),
+            Error::Base58(_) => f.write_str("base58 address encoding error"),
```
The rust way (at least how they define it in Error docs) is to either display inner error within display - or provide it as an output from `source`. One should not do both of these.

Kixunil replied:
Yes, I personally think returning error from source is better long-term.

> just felt a bit unintuitive not to give the concrete reason for the error.

Applications are supposed to iterate over error chain to display whole with their desired formatting (could be separated by `: `, newline, ...) Unfortunately there's no *convenient* way to do it now. AFIK Error WG works on it. I wrote some helpers in [random_stuff](https://github.com/Kixunil/rust_random_stuff) crate.

dr-orlovsky commented on src/util/address.rs:
```diff
@@ -482,24 +520,34 @@ impl Address {
+                    (WitnessVersion::V0, 32) => Ok(AddressType::P2wsh),
+                    (WitnessVersion::V0, len) => Err(Error::InvalidWitnessProgramLength(len)),
+                    (WitnessVersion::V1, 32) => Ok(AddressType::P2tr),
+                    (version, length) => Err(Error::UnsupportedWitnessVersion { version, length }),
```
The idea that the witness version is not supported by the library. It may be known to the software using the library

Kixunil replied:
Sounds like clarification in doc would be helpful.

dr-orlovsky commented:
@sgeisler thank you for the review! We have two ACKs from maintainers and one ACK from community, so I will merge it for now. But anyway lets finish the discussion of your nits in the comments and I will do another PR addressing them

dr-orlovsky commented:
Ok, we have version conflicts so re-review will be required anyway. I will address your nits with it - we just need to arrive to some end result with them

dr-orlovsky commented:
Rebased. @sgeisler I see your point with the `UnsupportedWitnessVersion` error, however since it requires more discussion and not a part of Taproot-specific update, I propose to leave it for a separate PR. It seems it is directly related to https://github.com/rust-bitcoin/rust-bitcoin/issues/648 discussion and we have to decide there on the overall API for `Address` type guarantees

Kixunil commented on src/util/address.rs:
```diff
@@ -71,30 +79,40 @@ pub enum Error {
-            Error::InvalidWitnessProgramLength(l) => write!(f,
-                "the witness program must be between 2 and 40 bytes in length: length={}", l,
+            Error::InvalidWitnessProgramLength(len) => write!(f,
+                "the witness program must be between 2 and 40 bytes in length: length={}", len,
```
One more general thought on error design, not related to this PR specifically: something like [`serde::de::Visitor::expecting`](https://docs.serde.rs/serde/de/trait.Visitor.html#required-methods) method or [`parse_arg::ParseArg::describe_type`](https://docs.rs/parse_arg/0.1.4/parse_arg/trait.ParseArg.html#required-methods) could be useful.

dr-orlovsky replied:
Not sure I entirely get what do you mean by that design

Kixunil replied:
Separate what is expected from what is the error. The intention is mainly to allow nicer/more flexible formatting of error messages.

Kixunil reviewed (approved):
utACK

sgeisler commented on src/util/address.rs:
```diff
@@ -71,30 +79,40 @@ pub enum Error {
     InvalidSegwitV0ProgramLength(usize),
     /// An uncompressed pubkey was used where it is not allowed.
     UncompressedPubkey,
+    /// Witness version which is not (yet) supported
```
I'd also like the docs to point out that wallets should still be able to send to the address in this case. Otherwise we'll end up with a big incompatibility mess again if people don't update their dependencies for whatever reason.

dr-orlovsky replied:
Added detailed error description

sgeisler commented on src/util/address.rs:
```diff
@@ -482,24 +520,34 @@ impl Address {
+                match (version, prog.len()) {
+                    (WitnessVersion::V0, 20) => Ok(AddressType::P2wpkh),
+                    (WitnessVersion::V0, 32) => Ok(AddressType::P2wsh),
+                    (WitnessVersion::V0, len) => Err(Error::InvalidWitnessProgramLength(len)),
```
Still relevant: why not `InvalidSegwitV0ProgramLength`? Seems to be made for exactly this case.

dr-orlovsky commented:
@Kixunil my point is that solving this issue will require more discussion and must involve many more changes from other PRs and issues (like #648, #642, #643, #646). The primary focus of this PR is to enable P2TR addresses _at the current master_, and we've being discussing and working on this trivial think already for 8 months having more than a hundred comments.

So while I am definitely up for a much better API for addresses, it is more generic that just support for P2TR proposed in this PR, and will require (potentially many) month(s) of discussion – so I propose to finally merge this one and move forward, unlocking a way for Taproot-supporting release.

Kixunil commented:
@dr-orlovsky Either I misclicked or GitHub have screwed up my comment, it was referencing line https://github.com/rust-bitcoin/rust-bitcoin/pull/563/files#diff-c55d0186f3d5990b5e223c1a8aa1bbe60988201c452dd13b3324e892bc33119eR88

I believe the documentation should say "but such address, while parsed, may return errors from some methods". (Parsing == converting string to internal representation)

This is unrelated to future designs as the documentation should reflect the state of things in code as written right now.

dr-orlovsky commented:
@Kixunil updated both error docs and display with the most verbose form I was able to compose. Pls review

Kixunil commented:
I mean, I specifically have issue with the word "**parsing**", which means [converting string to internal representation](https://en.wikipedia.org/wiki/Parsing). The literal `parse()` method is implemented on `&str` via [`FromStr`](https://github.com/rust-bitcoin/rust-bitcoin/pull/563/files#diff-c55d0186f3d5990b5e223c1a8aa1bbe60988201c452dd13b3324e892bc33119eR711) but that impl **does not** return `FutureWitnessVersion`. There's no other **parsing** method. All other methods do various other processing, not parsing. So the way the doc is written implies that such error is returned from `from_str`/`parse` which is not the case. It's the other way around - parsing does **not** return that variant, but other methods do.

apoelstra commented on src/util/address.rs:
```diff
@@ -344,6 +344,13 @@ impl From<WitnessVersion> for ::bech32::u5 {
+        version.bech32_variant()
+    }
+}
+
```
I don't like this commit. It seems to violate the spirit of `From`.

Kixunil replied:
Is it because you find those types unrelated? That's what I feel but curious if you have another reason.

apoelstra replied:
Yeah, I think that's what it is. It seems like this is changing, conceptually, what the data is.

apoelstra commented on src/util/address.rs:
```diff
@@ -531,23 +548,21 @@ impl Address {
+                    (WitnessVersion::V0, 32) => Ok(AddressType::P2wsh),
+                    (WitnessVersion::V0, len) => Err(Error::InvalidWitnessProgramLength(len)),
+                    (WitnessVersion::V1, 32) => Ok(AddressType::P2tr),
+                    (version, length) => Err(Error::FutureWitnessVariant { version, program_len: length }),
```
Why is this an error? I don't like this commit either.

Kixunil replied:
In the context of address *type* it does make sense (we don't know the address type) but maybe it enourages users to reject future versions, so should be a variant of `AddressType` instead?

dr-orlovsky replied:
@apoelstra you can't make it not a error unless we go further than just enabling P2TR addresses and will modify `AddressType` to support arbitrary future softforks - which will be a case for debate, already started in #648, #642, #643, #646...

Pls also see my explanations in https://github.com/rust-bitcoin/rust-bitcoin/pull/563#issuecomment-922424462

dr-orlovsky commented:
@apoelstra, I have removed both controversial commits and made this PR to be only about P2TR address support. So now:
- no `From<WitnessVersion> for bech32::u5`
- `address_type` reverted to return `Option` and not `Result`, as was before

The support of future Witness versions is independent from P2TR addresses and will be addressed in a separate PR(s), together with the rest of required address refactoring (#648, #642, #643, #646...).

apoelstra reviewed (approved):
ACK c1991d748f44730142b9978b80495d1d0e2d2e43

sanket1729 reviewed (approved):
crACK c1991d748f44730142b9978b80495d1d0e2d2e43

dr-orlovsky commented:
@Kixunil eaiting for you approval to merge this. I have removed all controversial part which was not directly related to P2TR address

Kixunil reviewed (approved):
Sorry for holding up, didn't make my mind when I saw your change and didn't know you want my review too.

ACK from me for this lighter change, I look forward into resolving the questions about invariants in the future.
